crates/matrix-sdk-common/src/ttl_cache.rs

```diff
@@ -0,0 +1,148 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! A TTL cache which can be used to time out repeated operations that might
+//! experience intermittent failures.
+
+use std::{borrow::Borrow, collections::HashMap, hash::Hash, time::Duration};
+
+use ruma::time::Instant;
+
+const DEFAULT_LIFETIME: Duration = Duration::from_secs(15 * 60);
+
+#[derive(Debug)]
+struct TtlItem<V: Clone> {
+    value: V,
+    insertion_time: Instant,
+    lifetime: Duration,
+}
+
+impl<V: Clone> TtlItem<V> {
+    fn expired(&self) -> bool {
+        self.insertion_time.elapsed() >= self.lifetime
+    }
+}
+
+/// A TTL cache where items get removed deterministically in the `get()` call.
+#[derive(Debug)]
+pub struct TtlCache<K: Eq + Hash, V: Clone> {
+    lifetime: Duration,
+    items: HashMap<K, TtlItem<V>>,
+}
+
+impl<K, V> TtlCache<K, V>
+where
+    K: Eq + Hash,
+    V: Clone,
+{
+    /// Create a new, empty, [`TtlCache`].
+    pub fn new() -> Self {
+        Self { items: Default::default(), lifetime: DEFAULT_LIFETIME }
+    }
+
+    /// Does the cache contain an non-expired item with the matching key.
+    pub fn contains<Q>(&self, key: &Q) -> bool
+    where
+        K: Borrow<Q>,
+        Q: Hash + Eq + ?Sized,
+    {
+        let cache = &self.items;
+        let contains = if let Some(item) = cache.get(key) { !item.expired() } else { false };
+
+        contains
+    }
+
+    /// Add a single item to the cache.
+    pub fn insert(&mut self, key: K, value: V) {
+        self.extend([(key, value)]);
+    }
+
+    /// Extend the cache with the given iterator of items.
+    pub fn extend(&mut self, iterator: impl IntoIterator<Item = (K, V)>) {
+        let cache = &mut self.items;
+
+        let now = Instant::now();
+
+        for (key, value) in iterator {
+            let item = TtlItem { value, insertion_time: now, lifetime: self.lifetime };
+
+            cache.insert(key, item);
+        }
+    }
+
+    /// Remove the item that matches the given key.
+    pub fn remove<Q>(&mut self, key: &Q) -> Option<V>
+    where
+        K: Borrow<Q>,
+        Q: Hash + Eq + ?Sized,
+    {
+        self.items.remove(key.borrow()).map(|item| item.value)
+    }
+
+    /// Get the item that matches the given key, if the item has expired `None`
+    /// will be returned and the item will be evicted from the cache.
+    pub fn get<Q>(&mut self, key: &Q) -> Option<V>
+    where
+        K: Borrow<Q>,
+        Q: Hash + Eq + ?Sized,
+    {
+        // Remove all expired items.
+        self.items.retain(|_, value| !value.expired());
+        // Now get the wanted item.
+        self.items.get(key.borrow()).map(|item| item.value.clone())
+    }
+
+    /// Force the expiry of the given item, if it is present in the cache.
+    ///
+    /// This doesn't remove the item, it just marks it as expired.
```

Comment:
Why not remove it immediately? If it's expired, is there any point in keeping it in the mapping?

Reply:
That method is for testing purposes mostly, perhaps I should make it`cfg(test)`.

It expires the item and then we check that a `get()` correctly removes it.

Reply:
Fair!